### DISCUSSION TITLE
localization (number, date, etc)

### DISCUSSION THREAD

**Discussion by willfarrell** [json-schema-org/community#242] [General]
I've been thinking a lot around creating data standards, defined by JSON Schema - data stored in csv, that are used by multiple locales. Creating multiple schema with different patterns and external code to transform values into `en-US` or `ISO` to be validated against the schema is the current approach, but would be nice if the spec was more inclusive. Just want to start the discussion around what it could look like in the spec, if this is best handled at the implementation layer, and/or what are best practices people are using.

## Intl.NumberFormat
Currently `type:"number"` accepts `1234.56` and if coercion is supported, "1234.56"

Numbers can also be written like "1,234.56", "1 234.56", "1.234,56", "1 234,56", depending on the locale. Could there be a new `format` to support validating these number formats based on a locale, passed in as an option or as part of the spec?

Ref: https://github.com/ajv-validator/ajv/issues/2099

## Intl.DateTimeFormat
Same as Intl.NumberFormat, but for dates. In this case the ISO standard is used, and not `en-US` which I think is the right approach, but should be considered for consistency.

## locale modifier
Ability to choose a what sub-schema to enforce based on the locale of the data. An example would be `enum` lists in different languages.

```json
{
  // similar to `not`?
  "locale:fr-CA": {
    ...
  }
}
```

For context, my locale is `en-CA`.

**Comment by gregsdennis** [comment]
The numeric format is specified by JSON, which is not something that we can alter.  Further (and [discussed elsewhere](https://github.com/json-schema-org/json-schema-vocabularies/issues/45)), numbers as strings is not something that we currently support.

In regard to the rest, for numerous reasons, we're recommending against using and/or expanding `format`.  It's touched on in that issue I linked above.  Essentially, these would best be supported as a vocabulary.

Lastly, there are many other issues/discussions that relate to the topic of localization:

- https://github.com/json-schema-org/json-schema-spec/issues/1241
- https://github.com/json-schema-org/json-schema-spec/issues/148
- https://github.com/json-schema-org/json-schema-spec/issues/278
- https://github.com/json-schema-org/json-schema-spec/issues/62
- https://github.com/json-schema-org/json-schema-spec/pull/12

**Comment by willfarrell** [comment]
Thanks so much for the resources, looks like I have some reading to do. Didn't think to look at the historical issues, sorry about the re-post.

**Comment by karenetheridge** [comment]
`1234.56` is a number, represented numerically by the system.

`"1,234.56"`, `"1 234.56"`, `"1.234,56"`, `"1 234,56"` are strings, that could be _converted_ to a number using the system's locale setting.

If the schema specifies a number, then it expects a number; when json-serialized it will always appear as `1234.56` (see json.org).

If you want to instead transmit a localised string that represents a number, you could declare a definition (under `$defs`) that indicated the locale format you want to use, e.g.:

```yaml
$defs:
  fr_FR_number:
    example: 1 234 567,89
    type: string
    pattern: ^\d{1,3}( \d{3})*,\d\d$
```

No doubt somewhere in the locale libraries bundled with the operating system there are regular expressions that could be imported in bulk and converted to json schema definitions.
